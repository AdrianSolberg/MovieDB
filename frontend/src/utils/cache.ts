import { makeVar } from "@apollo/client";
import { Filters, SortingType } from "@/types/movieTypes";

export const sortOptionVar = makeVar<SortingType>(
    (sessionStorage.getItem("sort_option") as SortingType) ??
        SortingType.NEWEST_FIRST
);


const storedFilters = sessionStorage.getItem("filters");
const initialFilters: Filters = storedFilters ? JSON.parse(storedFilters) : {
    Genre: [],
    Rating: [],
    Decade: [],
    Status: [],
    Runtime: [],
};
export const filtersVar = makeVar<Filters>(initialFilters);

export const searchVar = makeVar<string>(
    sessionStorage.getItem("search") ?? ""
);

export const usernameVar = makeVar<string>(
    localStorage.getItem("username") ?? "Guest"
);
