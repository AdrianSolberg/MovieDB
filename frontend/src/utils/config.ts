const BACKEND_PORT = 3001;

const BACKEND_URL =
    import.meta.env.MODE === "production"
        ? `http://it2810-26.idi.ntnu.no:${BACKEND_PORT}`
        : `http://localhost:${BACKEND_PORT}`;

export { BACKEND_URL };
