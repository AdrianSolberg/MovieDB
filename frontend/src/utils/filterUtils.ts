import { Filters, FiltersInput } from "@/types/__generated__/types";

export function getFiltersAsInput(filters: Filters): FiltersInput {
    return {
        Decade: filters.Decade.map((filter) => filter.name),
        Genre: filters.Genre.map((filter) => filter.name),
        Rating: filters.Rating.map((filter) => filter.name),
        Status: filters.Status.map((filter) => filter.name),
        Runtime: filters.Runtime.map((filter) => filter.name),
    };
}
