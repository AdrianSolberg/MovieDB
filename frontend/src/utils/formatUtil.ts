export const formatDate = (date: Date) => {
    return date.toLocaleDateString("en-US", {
        year: "numeric",
        month: "long",
        day: "numeric",
        hour: "2-digit",
        minute: "2-digit",
    });
};

export const formatNumber = (n: number) => {
    return n.toLocaleString("no-NO");
};

export const formatCurrency = (value: number) => {
    return value.toLocaleString("en-US", {
        style: "currency",
        currency: "USD",
        maximumFractionDigits: 0,
    });
};
