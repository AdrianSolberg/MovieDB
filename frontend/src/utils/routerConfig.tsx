import Navbar from "@/components/Navbar";
import ActivityPage from "@/pages/ActivityPage";
import HomePage from "@/pages/HomePage";
import MovieDetailPage from "@/pages/MovieDetailPage";
import MyReviewsPage from "@/pages/MyReviewsPage";
import NotFoundPage from "@/pages/NotFoundPage";
import { Toaster } from "@/shadcn/components/ui/sonner";
import { Outlet, RouteObject } from "react-router-dom";

const routerConfig: RouteObject[] = [
    {
        path: "/",
        element: (
            <>
                <Navbar />
                <div id="main-content" tabIndex={-1}>
                    <Outlet />
                </div>
                <Toaster position="bottom-right" />
            </>
        ),
        errorElement: <NotFoundPage />,
        children: [
            {
                path: "",
                element: <HomePage />,
            },
            {
                path: "movie/:movieId",
                element: <MovieDetailPage />,
            },
            {
                path: "activity",
                element: <ActivityPage />,
            },
            {
                path: "myReviews",
                element: <MyReviewsPage />,
            },
        ],
    },
];

export default routerConfig;
