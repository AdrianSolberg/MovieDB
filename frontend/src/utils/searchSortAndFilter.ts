import { Movie, SortingType } from "@/__generated__/types";

const filterByGenres = (movies: Movie[], selectedGenres: string[]) => {
    if (selectedGenres.length) {
        return movies.filter((movie) =>
            selectedGenres.every((selectedGenre) =>
                movie.genres.some((genre) => genre === selectedGenre)
            )
        );
    }
    return movies;
};

const filterByRating = (movies: Movie[], selectedRatings: string[]) => {
    if (selectedRatings.length) {
        return movies.filter((movie) =>
            selectedRatings.includes(
                Math.round(movie.vote_average / 2).toString()
            )
        );
    }
    return movies;
};

const filterByReleaseYear = (movies: Movie[], selectedDecades: string[]) => {
    if (selectedDecades.length) {
        return movies.filter((movie) => {
            const releaseYear = movie.release_date.getFullYear();
            return selectedDecades.some((decade) => {
                const startYear = parseInt(decade.slice(0, 4));
                return releaseYear >= startYear && releaseYear < startYear + 10;
            });
        });
    }
    return movies;
};

const filterByLanguage = (movies: Movie[], selectedLanguages: string[]) => {
    if (selectedLanguages.length) {
        return movies.filter((movie) =>
            selectedLanguages.every((selectedLanguage) =>
                movie.spoken_languages.some(
                    (language) => language === selectedLanguage
                )
            )
        );
    }
    return movies;
};

const filterByStatus = (movies: Movie[], selectedStatuses: string[]) => {
    if (selectedStatuses.length) {
        return movies.filter((movie) =>
            selectedStatuses.includes(movie.status)
        );
    }
    return movies;
};

const filterByRuntime = (movies: Movie[], selectedRuntimes: string[]) => {
    if (selectedRuntimes.length) {
        return movies.filter((movie) => {
            const runtime = movie.runtime;
            return selectedRuntimes.some((runtimeFilter) => {
                switch (runtimeFilter) {
                    case "Less than 1 hour":
                        return runtime < 60;
                    case "1 - 2 hours":
                        return runtime >= 60 && runtime < 120;
                    case "2 - 3 hours":
                        return runtime >= 120 && runtime < 180;
                    case "3 hours or more":
                        return runtime >= 180;
                    default:
                        return false;
                }
            });
        });
    }
    return movies;
};

export const filterMovies = (
    movies: Movie[],
    filters: { [key: string]: string[] }
) => {
    let filteredMovies = movies;

    const filterFunctions = {
        Genre: filterByGenres,
        Rating: filterByRating,
        "Release Year": filterByReleaseYear,
        Language: filterByLanguage,
        Status: filterByStatus,
        Runtime: filterByRuntime,
    };

    for (const [filterKey, filterFunction] of Object.entries(filterFunctions)) {
        filteredMovies = filterFunction(
            filteredMovies,
            filters[filterKey] || []
        );
    }

    return filteredMovies;
};

const sortingMap: Map<SortingType, (a: Movie, b: Movie) => number> = new Map<
    SortingType,
    (a: Movie, b: Movie) => number
>([
    [
        SortingType.NewestFirst,
        (a, b) => b.release_date.getTime() - a.release_date.getTime(),
    ],
    [
        SortingType.OldestFirst,
        (a, b) => a.release_date.getTime() - b.release_date.getTime(),
    ],
    [SortingType.BestRated, (a, b) => b.vote_average - a.vote_average],
    [SortingType.WorstRated, (a, b) => a.vote_average - b.vote_average],
    [SortingType.LongestRuntime, (a, b) => b.runtime - a.runtime],
    [SortingType.ShortestRuntime, (a, b) => a.runtime - b.runtime],
]);

export const sortMovies = (sortOption: SortingType, movies: Movie[]) => {
    if (sortingMap.has(sortOption)) {
        movies.sort(sortingMap.get(sortOption));
    }

    return movies;
};

export const searchMovies = (search: string, movies: Movie[]) => {
    return movies.filter((movie) =>
        movie.title.toLowerCase().includes(search.toLowerCase())
    );
};

const sortingDisplayNames: Map<SortingType, string> = new Map<
    SortingType,
    string
>([
    [SortingType.BestRated, "Best rated"],
    [SortingType.LongestRuntime, "Longest runtime"],
    [SortingType.NewestFirst, "Newest first"],
    [SortingType.OldestFirst, "Oldest first"],
    [SortingType.ShortestRuntime, "Shortest runtime"],
    [SortingType.WorstRated, "Worst rated"],
]);

export const getSortOptionDisplayName = (sortType: SortingType) => {
    return sortingDisplayNames.get(sortType);
};
