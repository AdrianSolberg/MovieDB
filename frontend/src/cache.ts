import { makeVar } from "@apollo/client";
import { SortingType } from "./utils/searchSortAndFilter";

export const sortOptionVar = makeVar<SortingType>(
    (sessionStorage.getItem("sort_option") as SortingType) ??
        SortingType.NEWEST_FIRST
);

export const filtersVar = makeVar<{ [key: string]: string[] }>(
    JSON.parse(sessionStorage.getItem("filters") ?? "{}")
);

export const searchVar = makeVar<string>(
    sessionStorage.getItem("search") ?? ""
);
