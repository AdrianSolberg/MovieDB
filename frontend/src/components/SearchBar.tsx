import { searchVar } from "@/cache";
import { Button } from "@/shadcn/components/ui/button";
import { Input } from "@/shadcn/components/ui/input";
import { Search } from "lucide-react";
import React, { useEffect, useState } from "react";

const SearchBar: React.FC = () => {
    const [search, setSearch] = useState<string>("");

    useEffect(() => {
        setSearch(searchVar());
    }, []);

    const handleSearch = (searchString: string) => {
        sessionStorage.setItem("search", searchString);
        searchVar(searchString);
    };

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setSearch(e.target.value);
        if (!e.target.value) {
            handleSearch(e.target.value.trim());
        }
    };

    const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
        const target = e.target as HTMLInputElement;
        if (e.key === "Enter" && target.value.trim()) {
            handleSearch(target.value.trim());
        }
    };

    return (
        <section className="flex flex-row flex-grow gap-2">
            <Input
                id="searchbar"
                name="searchbar"
                type="search"
                placeholder="Search..."
                value={search}
                onChange={handleChange}
                onKeyDown={handleKeyDown}
                aria-label="Search for movies"
            />
            <Button
                type="submit"
                onClick={() => handleSearch(search.trim())}
                aria-label="Search"
            >
                <Search />
            </Button>
        </section>
    );
};

export default SearchBar;
