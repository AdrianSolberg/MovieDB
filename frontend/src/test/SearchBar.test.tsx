import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { searchVar } from "@/utils/cache";
import SearchBar from "@/components/SearchBar";
import { vi } from "vitest";
import "@testing-library/jest-dom";

vi.mock("@/utils/cache", () => ({
    searchVar: vi.fn(),
}));

describe("SearchBar", () => {
    afterEach(() => {
        sessionStorage.clear();
    });

    it("renders the search input", () => {
        render(<SearchBar />);

        expect(screen.getByRole("searchbox")).toBeInTheDocument();
    });

    it("updates searchVar and sessionStorage when typing in the input field", async () => {
        render(<SearchBar />);

        const input = screen.getByRole("searchbox");
        await userEvent.type(input, "Inception");
        await new Promise((r) => setTimeout(r, 520));

        expect(searchVar).toHaveBeenCalledWith("Inception");
        expect(sessionStorage.getItem("search")).toBe("Inception");
    });

    it("trims input before setting searchVar and sessionStorage", async () => {
        render(<SearchBar />);

        const input = screen.getByRole("searchbox");
        await userEvent.type(input, " Inception   ");
        await new Promise((r) => setTimeout(r, 520));

        expect(searchVar).toHaveBeenCalledWith("Inception");
        expect(sessionStorage.getItem("search")).toBe("Inception");
    });

    it("flushes debounce and triggers search on Enter key", async () => {
        render(<SearchBar />);

        const input = screen.getByRole("searchbox");
        await userEvent.type(input, "Interstellar{enter}");

        expect(searchVar).toHaveBeenCalledWith("Interstellar");
        expect(sessionStorage.getItem("search")).toBe("Interstellar");
    });

    it("loads the initial search value from sessionStorage", () => {
        sessionStorage.setItem("search", "Avatar");

        render(<SearchBar />);

        expect(screen.getByRole("searchbox")).toHaveValue("Avatar");
    });

    it("clears search on empty input", async () => {
        render(<SearchBar />);

        const input = screen.getByRole("searchbox");
        await userEvent.type(input, "Dunkirk");
        await new Promise((r) => setTimeout(r, 520));

        await userEvent.clear(input);
        await new Promise((r) => setTimeout(r, 520));

        expect(searchVar).toHaveBeenCalledWith("");
        expect(sessionStorage.getItem("search")).toBe("");
    });
});
