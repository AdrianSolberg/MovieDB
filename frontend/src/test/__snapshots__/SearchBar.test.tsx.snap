// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="flex flex-grow flex-row gap-2"
  >
    <input
      aria-label="Search for movies"
      class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
      id="searchbar"
      maxlength="100"
      name="searchbar"
      placeholder="Search... (Typing updates result)"
      type="search"
    />
  </section>
</DocumentFragment>
`;
